crates/floresta-wire/src/p2p_wire/transport.rs

```diff
@@ -0,0 +1,353 @@
+    #[error("V2 protocol error: {0}")]
+    Protocol(#[from] ProtocolError),
+    #[error("V2 serde error: {0}")]
+    SerdeV2(#[from] bip324::serde::Error),
+    #[error("V1 serde error: {0}")]
+    SerdeV1(#[from] bitcoin::consensus::encode::Error),
+    #[error("Proxy error: {0}")]
+    Proxy(#[from] Socks5Error),
+}
+
+/// UTreeXO p2p message extensions to the base bitcoin protocol.
+pub enum UtreexoMessage {
+    Standard(NetworkMessage),
+    Block(UtreexoBlock),
+}
+
+pub enum ReadTransport<R: AsyncRead + Unpin + Send> {
+    V2(R, AsyncProtocolReader),
+    V1(R),
+}
+
+pub enum WriteTransport<W: AsyncWrite + Unpin + Send + Sync> {
+    V2(W, AsyncProtocolWriter),
+    V1(W, Network),
+}
+
+struct P2PMessageHeader {
+    _magic: Magic,
+    _command: [u8; 12],
+    length: u32,
+    _checksum: u32,
+}
+
+impl Decodable for P2PMessageHeader {
+    fn consensus_decode<R: bitcoin::io::Read + ?Sized>(
+        reader: &mut R,
+    ) -> std::result::Result<Self, bitcoin::consensus::encode::Error> {
+        let _magic = Magic::consensus_decode(reader)?;
+        let _command = <[u8; 12]>::consensus_decode(reader)?;
+        let length = u32::consensus_decode(reader)?;
+        let _checksum = u32::consensus_decode(reader)?;
+        Ok(Self {
+            _checksum,
+            _command,
+            length,
+            _magic,
+        })
+    }
+}
+
+/// Establishes a TCP connection and negotiates the bitcoin protocol.
+///
+/// This function tries to connect to the specified address and negotiate the bitcoin protocol
+/// with the remote node. It first attempts to use the V2 protocol, and if that fails with a specific
+/// error suggesting fallback to V1 protocol (and `allow_v1_fallback` is true), it will retry
+/// the connection with the V1 protocol.
+///
+/// # Arguments
+///
+/// * `address` - The address of a target node
+/// * `network` - The bitcoin network
+/// * `allow_v1_fallback` - Whether to allow fallback to V1 protocol if V2 negotiation fails
+///
+/// # Returns
+///
+/// Returns a tuple of read and write transports that can be used to communicate with the node.
+///
+/// # Errors
+///
+/// Returns a `TransportError` if the connection cannot be established or protocol negotiation fails.
+pub async fn connect<A: ToSocketAddrs>(
+    address: A,
+    network: Network,
+    allow_v1_fallback: bool,
+) -> Result<
+    (
+        ReadTransport<ReadHalf<TcpStream>>,
+        WriteTransport<WriteHalf<TcpStream>>,
+    ),
+    TransportError,
+> {
+    match try_connection(&address, network, false).await {
+        Ok(transport) => Ok(transport),
+        Err(TransportError::Protocol(ProtocolError::Io(_, ProtocolFailureSuggestion::RetryV1)))
+            if allow_v1_fallback =>
+        {
+            try_connection(&address, network, true).await
+        }
+        Err(e) => Err(e),
+    }
+}
+
+async fn try_connection<A: ToSocketAddrs>(
+    address: &A,
+    network: Network,
+    force_v1: bool,
+) -> Result<
```

Comment:
I think these can be turned into some `type` declaration. They are pretty verbose

Reply:
Hm yea that makes sense, I think I had some before with the trait bounds that ended up being more verbose so I tossed them all. But since these are more concrete maybe an alias would work?